### PR TITLE
ci: Refactor creation of the public PRs

### PR DESCRIPTION
**Pull request by Milica-Cvrkota-IBM**
Changes:
* extract common scripts
* added a new task that runs first:

     - Checks for ALL open PRs
     - Rebases all the PR branches and the forked main from the upstream main

* if instana-operator-{release} PR is open, don't create a new one/don't commit changes
* if it's not open, commit changes on the feature branch and open a PR

### CODE REVIEW

File: ci/scripts/commit-changes-to-public-repo.sh [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+    popd
+fi
+
+
+git config --global user.name "instanacd"
+git config --global user.email "instanacd@instana.com"
+
+git add .
+git commit -s -m "$COMMIT_MESSAGE" --allow-empty
```
Comment (konrad-ohms) [inline ci/scripts/commit-changes-to-public-repo.sh:76]:
```suggestion
git commit -s -m "$COMMIT_MESSAGE"
# use force to ensure updated branches
# potential optimization -> check if changes on the remote and local branch are equal on a file level to not push a new commit if not necessary (might trigger a new pipeline run)
# we could do that by trying to fetch the remote branch and apply the changes, if not files changed, do not push it
git push origin "$PR_BRANCH_NAME" --force
```

File: ci/scripts/create-pr.sh [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+popd
+
+# Create the PR
+set -x
+curl \
+-fX POST \
+-H "Accept: application/vnd.github+json" \
+-H "Authorization: Bearer $GH_API_TOKEN" \
+https://api.github.com/repos/$OWNER/$REPO/pulls \
+-d "{\"title\":\"$COMMIT_MESSAGE\",\"head\":\"instana:main\",\"base\":\"main\"}"
```
Comment (konrad-ohms) [inline ci/scripts/create-pr.sh:35]: Just a matter of styling, I prefer to intend the command flags if they are split into multiple lines, but no hard requirement.
But if we decide to use a feature branch, we need to adjust the PR creation accordingly.
```suggestion
curl \
  -fX POST \
  -H "Accept: application/vnd.github+json" \
  -H "Authorization: Bearer $GH_API_TOKEN" \
  https://api.github.com/repos/$OWNER/$REPO/pulls \
  -d "{\"title\":\"$COMMIT_MESSAGE\",\"head\":\"instana:$PR_BRANCH_NAME\",\"base\":\"main\"}"
```

**Review by konrad-ohms — CHANGES_REQUESTED** [review]
Thanks, looks good in general.
I have a few remarks, especially using feature branches instead of main might be useful

File: ci/scripts/commit-changes-to-public-repo.sh [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+# (c) Copyright Instana Inc.
+#
+
+set -e
+set -o pipefail
+
+abort_if_pr_exists() {
+    echo "Check if a PR is already open"
+    PR_LIST_JSON=$(curl -fH "Accept: application/vnd.github+json" https://api.github.com/repos/"$OWNER"/"$REPO"/pulls)
+    EXISTING_PR_INFO_JSON=$(echo "$PR_LIST_JSON" | jq ".[] | select(.title | contains(\"$OPERATOR_PUBLIC_PR_NAME\"))")
```
[the remark this answers is not in the document]
Comment (Milica-Cvrkota-IBM) [reply ci/scripts/commit-changes-to-public-repo.sh:14]: Hmm good question. In the past we had to manually create the PRs sometimes that were real PRs that should be considered for a release promotion. If we agree that we can't do this going forward, we can add the user ID to the check

File: ci/scripts/commit-changes-to-public-repo.sh [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    if [ -n "$EXISTING_PR_INFO_JSON" ]; then
+        echo "A PR is already open, exiting"
+        echo "A job can be retriggered once the open PR is resolved"
+        exit 0
+    fi
```
[the remark this answers is not in the document]
Comment (Milica-Cvrkota-IBM) [reply ci/scripts/commit-changes-to-public-repo.sh]: I'm not sure that the changes would change in the second iteration once the PR is already created with the new changes

File: ci/scripts/commit-changes-to-public-repo.sh [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+popd
+
+pushd "$PUBLIC_REPO_LOCAL_NAME"/operators/"$OPERATOR_NAME"
+
+echo "Rebasing the fork from the upstream"
+git remote add upstream https://github.com/"$OWNER"/"$REPO".git
+git fetch upstream
+git rebase upstream/main
+git push --force
```
[the remark this answers is not in the document]
Comment (Milica-Cvrkota-IBM) [reply ci/scripts/commit-changes-to-public-repo.sh]: We have a `git checkout` now

**Review by konrad-ohms — APPROVED** [review]
Looks good, thanks :slightly_smiling_face: 
I am wondering if we should add some body to the PRs the automation create, but it looks like it is not required by Red Hat, so just personal preference and we could not add much info anyways I think

**Pull request merged**